### PR TITLE
add --use-specter to desi_run_night

### PR DESCRIPTION
**Pull request by marcelo-alvarez**
Currently the choice of specter vs gpu_specter when running desi_run_night is dependent on system_name, such that gpu_specter is only used for system_name == 'perlmutter-gpu'. 

This PR makes gpu_specter the default for any value of system_name with desi_run_night. Specter is only used if --use-specter is specified. GPUs will be used with gpu_specter if system_name == 'perlmutter-gpu'.

I have checked that the slurm files are as expected by running desi_run_night --dry-run-level 1 ... on cori and perlmutter.

@sbailey and @akremin please review for style and unintended consequences. 

<s>
We can either (1) merge this PR as is or (2) change the default to gpu_specter by using --use-specter, instead, since eventually gpu_specter will be the default. 
</s>

### CODE REVIEW

File: py/desispec/workflow/procfuncs.py [hunk 1]
```diff
@@ -1158,7 +1164,7 @@ def submit_tilenight(ptable, prows, calibjobs, internal_id, queue, reservation,
 def science_joint_fit(ptable, sciences, internal_id, queue='realtime', reservation=None,
                       z_submit_types=None, dry_run=0, strictly_successful=False,
                       check_for_outputs=True, resubmit_partial_complete=True,
-                      system_name=None):
+                      system_name=None,use_gpuspecter=False):
```
Comment (akremin) [inline py/desispec/workflow/procfuncs.py:1167]: This doesn't seem to be passed on to the `joint_fit` function here (this is just a wrapper). You did add the hooks to `joint_fit`, so you just need to add the argument in the function call on line 1178.

**Review by akremin — COMMENTED** [review]
Thanks Marcelo. I made one comment inline where I think an additional argument propagation is needed. 

Given that we hope to be running with `gpuspecter` quite soon (ideally for Iron), I think it would be best to change this to opt-in to specter and have `gpuspecter` as default. I.e. option (2) with `--use-specter`. Apologies for making you adjust all your changes but I think it is a short timescale before we will be using `gpuspecter` and we generally want the default to be no new arguments.

File: py/desispec/workflow/procfuncs.py (hunk 1 above)
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py:1167]: Thanks for catching this. As I understand it, jobs processing multiple exposures (other than tilenights) don't involve the extraction step. So does anything called by `checkfor_and_submit_joint_job` even need to know about what kind of specter to use?  If not, I'll just get rid of `use_gpuspecter` in the functions `checkfor_and_submit_joint_job`, `science_joint_fit`, and `joint_fit`.

**Comment by marcelo-alvarez** [comment]
> Thanks Marcelo. I made one comment inline where I think an additional argument propagation is needed.
> 
> Given that we hope to be running with `gpuspecter` quite soon (ideally for Iron), I think it would be best to change this to opt-in to specter and have `gpuspecter` as default. I.e. option (2) with `--use-specter`. Apologies for making you adjust all your changes but I think it is a short timescale before we will be using `gpuspecter` and we generally want the default to be no new arguments.

@akremin No worries. I figured this would be how we end up doing it but wanted to get the ball rolling on the PR with the still-specter-as-default version I'm using for testing gpu_specter end-to-end on cori and perlmutter-cpu, in particular to get the review of passing the use_[gpu]specter boolean from function to calling function out of the way. I will switch `use_gpuspecter` to `use_specter` after I finish the last gpu_specter validation tests, in this PR, probably today or tomorrow.

File: py/desispec/workflow/procfuncs.py (hunk 1 above)
Comment (marcelo-alvarez) [reply py/desispec/workflow/procfuncs.py]: @akremin I went ahead and got rid of `use_specter` in the functions for "joint" fits and jobs, since they don't involve extractions. Please check this to make sure we are not missing it anywhere after this change.

**Comment by marcelo-alvarez** [comment]
@sbailey I have checked the latest commit with and without --use-specter with `desi_run_night --dry-run-level 1` on cori-haswell, perlmutter-cpu and perlmutter-gpu, with results as expected. Feel free to test yourself and/or merge.

**Comment by akremin** [comment]
Thank Marcelo. It is hard to parse the git diff here to verify it covers everything, but what I see looks good. I'll check some scripts tomorrow and ideally run an example or two to verify functionality.

**Comment by sbailey** [comment]
In `/global/cfs/cdirs/desi/users/sjbailey/spectro/redux/p8r` I tried running
```
desi_run_night -n 20210222 --z-submit-types pernight --use-tilenight -q regular --dry-run-level 1 --system-name perlmutter-gpu &> run/logs/20210222.log
```
it looks like the resulting tilenight jobs are good, with `desi_proc_tilenight ... --gpuspecter --gpuextract`, but the flat jobs run `desi_proc` without requesting `--gpuspecter --gpuextract`.

@marcelo-alvarez please check on that; the `desi_proc` calls for flats should use the same gpuspecter options as for science exposures.

Note: let's keep the scope of this PR to cover the cases when `desi_run_night --system-name` is specified.  After that is working, I'll rebase my branch and submit another PR to handle the cases where `--system-name` is not specified and it is auto-picking based upon job type (issue #1881).

**Comment by sbailey** [comment]
@marcelo-alvarez please also add `desi_daily_proc_manager --use-specter` option so that we can merge this but still continue to use specter for daily processing until the validation runs are completed and documented.

Otherwise the code updates and generated scripts look good, though I'm also doing a paranoia check of also running them for a night.

**Comment by marcelo-alvarez** [comment]
@sbailey flat jobs now run with `--gpuspecter` and `--gpuextract` where expected, and `desi_daily_proc_manager --use-specter` option has been added. We should make sure to test the latter in a realistic way, given that it could affect the daily processing.

**Comment by akremin** [comment]
I ran the `desi_daily_proc_manager` with both `--use-specter` and without, and it did the correct thing for all job types in both circumstances. Eventually we will want to change `--gpu-specter` in the lower level functions to also use (or not use) `--use-specter`, but that can be left for a future PR.

**Comment by akremin** [comment]
Since @marcelo-alvarez and @sbailey have tested the actually running of `specter` vs. `gpu-specter` and the `desi_run_night` implementation in this PR, I will now merge.
